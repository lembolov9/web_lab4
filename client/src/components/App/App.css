.App {
    max-width: 1000px;
    margin: 0 auto;
    background-color: lavender;
    font-family: 'Georgia', serif ;
    border-radius: 15px;
}

.header {
    margin: 20px;
    display: flex;
    justify-content: space-around;
}

.geoBtn {
    border-radius: 10px;
    margin-top: 20px ;
    background-color: thistle;
}


