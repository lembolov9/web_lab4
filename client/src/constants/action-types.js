export const ADD_CITY = 'ADD_CITY';
export const ADD_CITY_SUCCESS = 'ADD_CITY_SUCCESS';
export const ADD_CITY_ERROR = 'ADD_CITY_ERROR';
export const WEATHER_WAIT = 'WEATHER_WAIT';
export const WEATHER_SUCCESS = 'WEATHER_SUCCESS';
export const WEATHER_ERROR = 'WEATHER_ERROR';
export const DELETE_CITY = 'DELETE_CITY';
export const DELETE_CITY_SUCCESS = 'DELETE_CITY_SUCCESS';
export const DELETE_CITY_ERROR = 'DELETE_CITY_ERROR';
export const GET_CITIES_WAIT = 'GET_CITIES_WAIT';
export const GET_CITIES_SUCCESS = 'GET_CITIES_SUCCESS';
export const GET_CITIES_ERROR = 'GET_CITIES_ERROR';