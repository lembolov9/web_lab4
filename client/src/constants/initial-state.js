export const INITIAL_STATE = {
    cityList: {
        cities: [],
        newCity: null,
        error: null,
        isGetting: false,
        cityForDelete: null
    },
    mainCity: {
        id: 0
    }
};

export const DEFAULT = {
    'longitude': 37.62,
    'latitude': 55.75
};