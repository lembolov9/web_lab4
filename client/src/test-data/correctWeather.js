export default  {
    "coord": {
        "lon": 30.52,
        "lat": 50.43
    },
    "weather": [
        {
            "id": 500,
            "main": "Rain",
            "description": "light rain",
            "icon": "10n"
        }
    ],
    "base": "stations",
    "main": {
        "temp": 8.02,
        "pressure": 1011,
        "humidity": 100,
        "temp_min": 7.78,
        "temp_max": 8.33
    },
    "visibility": 6000,
    "wind": {
        "speed": 3,
        "deg": 80
    },
    "rain": {
        "1h": 0.76
    },
    "clouds": {
        "all": 90
    },
    "dt": 1573154160,
    "sys": {
        "type": 1,
        "id": 8903,
        "country": "UA",
        "sunrise": 1573102677,
        "sunset": 1573136713
    },
    "timezone": 7200,
    "id": 703448,
    "name": "Kiev",
    "cod": 200
};