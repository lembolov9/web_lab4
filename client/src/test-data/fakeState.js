export const FAKE_STATE = {
    'cityList': {
        'cities': [{"id":1,"name":"moscow"},{"id":2,"name":"kiev"},{"id":3,"name":"kiev"}],
        'newCityId': 4,
    },
    'mainCity': {
        'id': 0
    }
};