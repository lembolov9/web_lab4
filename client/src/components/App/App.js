import React from "react"
import Weather from "../Weather/Weather";
import axios from "axios";
import {DEFAULT} from "../../constants/initial-state";
import { connect } from 'react-redux';
import {getMainCity} from "../../reducers"
import {getWeatherError, getWeatherSuccess, getWeatherWait} from "../../actions";
import "./App.css"
import {WList} from "../WeatherList/WeatherList";
import {Add} from "../AddForm/AddForm";
import {api} from "../../constants/server";

const getWeather = (dispatch, coords, id) => {
    axios
        .get(`${api}weather/coordinates?lat=${coords.latitude}&long=${coords.longitude}`)
        .then(response => {
            dispatch(getWeatherSuccess(response.data, id));
        })
        .catch(error => {
            dispatch(getWeatherError(error, id));
        });
};


export const getWeatherByCoordsAction = () => {
    return dispatch => {
        const id = 0;
        const options = {
            enableHighAccuracy: true,
            timeout: 3000
        };

        dispatch(getWeatherWait(id));

        if (!navigator.geolocation) {
            getWeather(dispatch, DEFAULT, id);
            console.log("Geolocation is missing");
        } else {
            navigator.geolocation.getCurrentPosition((position) => {
                getWeather(dispatch, position.coords, id);
            }, err => {
                        getWeather(dispatch, DEFAULT, id);
                        console.log("Geolocation error: " + err);
                        }, options);
        }
    }
};


class App extends React.Component {
    render() {
        return (
                <div className="App">
                    <div className="header">
                        <div><h1>Weather</h1></div> <div><button className="geoBtn" onClick={() => this.props.getWeatherByCoords()}>Change Geolocation</button></div>
                    </div>
                    <Weather key={0} data={this.props.mainCity} getWeather={() => this.props.getWeatherByCoords()}/>
                    <Add/>
                    <WList/>
                </div>
        );
    }
}

const mapStateToProps = (state) => ({
    mainCity: getMainCity(state)
});

const mapDispatchToProps = (dispatch) => ({
    getWeatherByCoords: () => dispatch(getWeatherByCoordsAction())
});

const AppC = connect(mapStateToProps, mapDispatchToProps)(App);

export {AppC, App};