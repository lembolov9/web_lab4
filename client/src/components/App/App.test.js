import renderer from "react-test-renderer";
import configureMockStore from "redux-mock-store";
import React from "react";
import {App} from "./App";
import {Provider} from "react-redux";
import {INITIAL_STATE} from "../../constants/initial-state";
import correctWeather from "../../test-data/correctWeather";
import {FAKE_STATE} from "../../test-data/fakeState";
import thunk from "redux-thunk";

const mockStore = configureMockStore();
const fakeState = INITIAL_STATE;
fakeState.mainCity.weather = correctWeather;
fakeState.mainCity.waiting = false;

const fakeStore = mockStore(fakeState);

it("App display correctly without favourites", () => {
    const component = renderer.create(
        <Provider store={fakeStore}>
            <App getWeatherByCoords={() => {}} mainCity={fakeState.mainCity} />
        </Provider>
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});


const mockStoreWithFav = configureMockStore([thunk]);
const fakeStateWithFav = FAKE_STATE;
fakeState.mainCity.weather = correctWeather;
fakeState.mainCity.waiting = false;

const fakeStoreWithFav = mockStoreWithFav(fakeStateWithFav);

it("App display correctly with favourites", () => {
    const component = renderer.create(
        <Provider store={fakeStoreWithFav}>
            <App getWeatherByCoords={() => {}} mainCity={fakeState.mainCity} />
        </Provider>
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});