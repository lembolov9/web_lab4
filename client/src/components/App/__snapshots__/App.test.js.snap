// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App display correctly with favourites 1`] = `
<div
  className="App"
>
  <div
    className="header"
  >
    <div>
      <h1>
        Weather
      </h1>
    </div>
     
    <div>
      <button
        className="geoBtn"
        onClick={[Function]}
      >
        Change Geolocation
      </button>
    </div>
  </div>
  <div
    className="Weather grid"
  >
    <div
      className="mainInfo grid"
    >
      <div
        className="cityName"
      >
        <h3>
          Kiev
        </h3>
      </div>
      <div
        className="cityTemp"
      >
        8.02
         ℃
      </div>
      <div
        className="cityIcon"
      >
        <img
          alt="weather icon"
          src="https://openweathermap.org/themes/openweathermap/assets/vendor/owm/img/widgets/10n.png"
        />
      </div>
    </div>
    <div
      className="Info flex"
    >
      <div>
        Wind speed:
      </div>
      <div>
        3
         m/s
      </div>
    </div>
    <div
      className="propsRow flex"
    >
      <div>
        Clouds:
      </div>
      <div>
        90
        %
      </div>
    </div>
    <div
      className="propsRow flex"
    >
      <div>
        Pressure:
      </div>
      <div>
        1011
        hpa
      </div>
    </div>
    <div
      className="propsRow flex"
    >
      <div>
        Humidity:
      </div>
      <div>
        100
        %
      </div>
    </div>
    <div
      className="propsRow flex"
    >
      <div>
        Coordinates:
      </div>
      <div>
        [
        30.52
        , 
        50.43
        ]
      </div>
    </div>
  </div>
  <div
    className="AddForm"
  >
    <div>
      <h2>
        Favourites
      </h2>
    </div>
    <div>
      <form
        onSubmit={[Function]}
      >
        <input
          id="addCity"
          placeholder="Add city"
          required={true}
        />
        <button
          className="addBtn"
          type="submit"
        >
          +
        </button>
      </form>
    </div>
  </div>
  <div
    className="WeatherList"
  >
    <div
      className="Weather"
    >
      <div
        className="loader"
      />
    </div>
    <div
      className="Weather"
    >
      <div
        className="loader"
      />
    </div>
    <div
      className="Weather"
    >
      <div
        className="loader"
      />
    </div>
  </div>
</div>
`;

exports[`App display correctly without favourites 1`] = `
<div
  className="App"
>
  <div
    className="header"
  >
    <div>
      <h1>
        Weather
      </h1>
    </div>
     
    <div>
      <button
        className="geoBtn"
        onClick={[Function]}
      >
        Change Geolocation
      </button>
    </div>
  </div>
  <div
    className="Weather grid"
  >
    <div
      className="mainInfo grid"
    >
      <div
        className="cityName"
      >
        <h3>
          Kiev
        </h3>
      </div>
      <div
        className="cityTemp"
      >
        8.02
         ℃
      </div>
      <div
        className="cityIcon"
      >
        <img
          alt="weather icon"
          src="https://openweathermap.org/themes/openweathermap/assets/vendor/owm/img/widgets/10n.png"
        />
      </div>
    </div>
    <div
      className="Info flex"
    >
      <div>
        Wind speed:
      </div>
      <div>
        3
         m/s
      </div>
    </div>
    <div
      className="propsRow flex"
    >
      <div>
        Clouds:
      </div>
      <div>
        90
        %
      </div>
    </div>
    <div
      className="propsRow flex"
    >
      <div>
        Pressure:
      </div>
      <div>
        1011
        hpa
      </div>
    </div>
    <div
      className="propsRow flex"
    >
      <div>
        Humidity:
      </div>
      <div>
        100
        %
      </div>
    </div>
    <div
      className="propsRow flex"
    >
      <div>
        Coordinates:
      </div>
      <div>
        [
        30.52
        , 
        50.43
        ]
      </div>
    </div>
  </div>
  <div
    className="AddForm"
  >
    <div>
      <h2>
        Favourites
      </h2>
    </div>
    <div>
      <form
        onSubmit={[Function]}
      >
        <input
          id="addCity"
          placeholder="Add city"
          required={true}
        />
        <button
          className="addBtn"
          type="submit"
        >
          +
        </button>
      </form>
    </div>
  </div>
  <div
    className="WeatherList"
  />
</div>
`;
